ggml-metal.metal

```diff
@@ -729,12 +731,12 @@ kernel void kernel_rope(
         }
     } else {
         for (int64_t ib = 0; ib < ne0/n_dims; ++ib) {
-            for (int64_t ic = 0; ic < n_dims; ic += 2) {
+            for (int64_t ic = 2*tiitg; ic < n_dims; ic += 2*tptg.x) {
+
+                const float theta = theta_0 * pow(freq_base, inv_ndims*ic - ib);
```

Comment:
Could someone explain this line to me? Why isn't this just `pow(freq_base, inv_ndims*i0)`?

Reply:
I think the reason close to #3426

Reply:
`inv_ndims*ic - ib` ->
`ic/n_dims - (ib*n_dims)/n_dims` ->
`(ic - ib*n_dims)/n_dims`

I think the line should be:

```c
const float theta = theta_0 * pow(freq_base, inv_ndims*ic + ib);
```

Using:

```c
pow(freq_base, inv_ndims*i0)

Reply:
Is there a paper or reference implementation I can look at to help me understand what these values represent? The transformers implementation just uses the same RoPE as LLaMA, so that's no help.

Reply:
The RoPE is explained here: https://arxiv.org/pdf/2104.09864.pdf

The different modes in ggml correspond to different memory layout of the data to be roped.
You can consider the CPU implementation as reference:

https://github.com/ggerganov/llama.cpp/blob/d82a5c026991cb7be346cbc4dc2b8b0caeca09e5/ggml.c#L13475-L13495

Reply:
I thought about the CPU implementation some more and realized that I forgot a *2* because of the way theta_scale is written. Here's what makes sense to me:

$$\Large{\theta = p \cdot freq\\_base ^ {-2 i_0 / n\\_dims}}$$

In the CPU implementation, for each inner loop, with a count that can be determined by half of *ic* plus *n_dims* for every *ib* ($`ib \cdot n\_dims + ic / 2`$), theta is multiplied by theta_scale. Since theta_scale is doubled, we must double the exponent here too.

The Metal implementation expands this to the point of being unrecognizable:

$$\Large{\theta = p \cdot freq\\_base ^ {-2 (ib \cdot n\\_dims + ic/2) / n\\_dims}}$$

$$\Large{\theta = p \cdot freq\\_base ^ {-2 (ib + ic/(2 \cdot n\\_dims) )}}$$

$$\Large{\theta = p \cdot freq\\_base ^ { ic/n\\_dims - 2 \cdot ib }}$$

$$\Large{\theta = p \cdot freq\\_base ^ { ic(-1/n\\_dims) - 2 \cdot ib }}$$

$$\Large{\theta = p \cdot freq\\_base ^ {inv\\_ndims \cdot ic - 2 \cdot ib }}$$

But where is the *2* in the Metal version? Based on the comment, it sounds like this code is usually run with *ne0* == *n_dims*, so *ib* is usually zero, and this may not make a numerical difference in practice. But since they don't seem to agree, I'm not sure how to extend them.

Reply:
> In the CPU implementation

For each value of `ib`, we perform `n_dims/2` multiplications with `theta_scale` because `ic` steps with `+2`, while you assume `n_dims` multiplications. So I think you should not use `i0 = ib*n_dims + ic/2`, but `i1 = ib*(n_dims/2) + ic/2` in the exponent

Reply:
Okay, now it makes sense to me. Thanks.

Reply:
@ggerganov Should I make a PR that changes the CUDA implementation to match CPU and Metal?

Reply:
Yup, it's missing the `ib` component. I'm actually still not sure about it as per the comment in the CPU implementation. I just think we've never ran a model that has `n_rot` less than `n_dims`.

But for consistency sake, we should add it to CUDA as well.